### ISSUE TITLE
Operator to ensure an expression is contextually typed by, and satisfies, some type

### ISSUE THREAD

**Issue by magnushiie** [microsoft/TypeScript#7481]
Sometimes it's necessary (e.g. for guiding type inference, for ensuring sub-expression conforms to an interface, or for clarity) to change the static type of an expression. Currently TypeScript has the `as` (aka `<>`) operator for that, but it's dangerous, as it also allows down-casting. It would be nice if there was another operator for implicit conversions only (type compatibility). I think this operator should be recommended in most cases instead of `as`.

This operator can be implemented as a generic function, but as it shouldn't have any run-time effect, it would be better if it was incorporated into the language.

``` typescript
function asType<T>(value: T) {
  return value;
};
```

EDIT: Due to parameter bivariance, this function is _not_ equivalent to the proposed operator, because asType allows downcasts too.

**Comment by RyanCavanaugh** [comment]
Can you post a few examples of how you'd like this to work so we can understand the use cases?

**Comment by magnushiie** [comment]
One example very close to the real-world need (I'm trying to get react and react-redux typings to correctly represent the required/provided Props):
```typescript
import { Component } from "react";
import { connect } from "react-redux";

// the proposed operator, implemented as a generic function
function asType<T>(value: T) {
  return value;
};

// in real life, imported from another (actions) module
function selectSomething(id: string): Promise<void> {
  // ...
  return null;
}

interface MyComponentActions {
  selectSomething(id: string): void;
}

class MyComponent extends Component<MyComponentActions, void> {
  render() {
    return null;
  }
}

// I've changed the connect() typing from DefinitelyTyped to the following:
// export function connect<P, A>(mapStateToProps?: MapStateToProps,
//                            mapDispatchToProps?: MapDispatchToPropsFunction|A,
//                            mergeProps?: MergeProps,
//                            options?: Options): ComponentConstructDecorator<P & A>;

// fails with "Argument of type 'typeof MyComponent' not assignable" because of 
// void/Promise<void> mismatch - type inference needs help to upcast the expression
// to the right interface so it matches MyComponent
export const ConnectedPlain = connect(undefined, {
  selectSomething,
})(MyComponent);

// erronously accepted, the intention was to provide all required actions
export const ConnectedAs = connect(undefined, {
} as MyComponentActions)(MyComponent);

// verbose, namespace pollution
const actions: MyComponentActions = {
  selectSomething,
};
export const ConnectedVariable = connect(undefined, actions)(MyComponent);

// using asType<T>(), a bit verbose, runtime overhead, but otherwise correctly verifies the
// expression is compatible with the type
export const ConnectedAsType = connect(undefined, asType<MyComponentActions>({
  selectSomething,
}))(MyComponent);

// using the proposed operator, equivalent to asType, does not compile yet
export const ConnectedOperator = connect(undefined, {
  selectSomething,
} is MyComponentActions)(MyComponent);
```

I've called the proposed operator in the last snippet ```is```.

The other kind of scenario is complex expressions where it's not immediately obvious what the type of the expression is and helps the reader understand the code, and the writer to get better error messages by validating the subexpression types individually. This is especially useful in cases of functional arrow function expressions.

A somewhat contrived example (using the tentative ```is``` operator again), where it's not immediately obvious what the result of getWork is, especially when it's a generic function where the result type depends on the argument type:
```typescript
const incompleteTasks = (tasks: Task[]) => tasks.filter(task => !(getWork(task.currentAssignment) is Todo).isComplete);
```

**Comment by DanielRosenwasser** [comment]
I ran into something similar when I was patching up code in DefinitelyTyped - to get around checking for excess object literal assignment, you have to assert its type, but that can be a little extreme in some circumstances, and hides potential issues you might run into during a refactoring.

There are also scenarios where I want to "bless" an expression with a contextual type, but I don't want a full blown type assertion for the reasons listed above. For instance, if a library defines a type alias for its callback type, I want to contextually type my callback, but I ***don't*** want to use a type assertion.

In other words, a type assertion is for saying "I know what I'm going to do, leave me a alone." This is more for "I'm pretty sure this should be okay, but please back me up on this TypeScript".

**Comment by RyanCavanaugh** [comment]
Sounds a lot like #2876?

**Comment by magnushiie** [comment]
If I understand #2876 correctly, it's still a downcast (i.e. bypassing type safety). What I was proposing here is an upcast (i.e. guaranteed to succeed at runtime or results in compile time error). Also, while ```<?>``` seems a bit like magic, the ```is``` operator is as straightforward as assigning to a variable with a defined type or passing an argument to a function with a parameter that has a defined type.

I think the best example of this operator exists in the Coq language:
```coq
Definition id {T} (x: T) := x. 
Definition id_nat x := id (x : nat).
Check id_nat.
id_nat
     : nat -> nat
```

Here, the expression ```x : nat``` is a [type cast](https://coq.inria.fr/refman/Reference-Manual003.html#typecast), where Coq's type cast is not dynamic but static (and mostly used in generic scenarios, like the ones I mentioned above) - here it means id_nat's argument type is restricted to be a nat.

**Comment by chilversc** [comment]
Another case for this is when returning an object literal from a function that has a type union for it's return type such as `Promise.then`.

``` typescript
interface FeatureCollection {
  type: 'FeatureCollection'
  features: any[];
}

fetch(data)
  .then(response => response.json())
  .then(results => ({ type: 'FeatureCollection', features: results }));
```

This gets quite tricky for intellisense in VS because the return type from `then` is `PromiseLike<T> | T`. Casting allows intellisense to work, but as mentioned it can hide errors due to missing members.

Also the error messages when the return value is invalid are quite obtuse because they refer to the union type. Knowing the intended type would allow the compiler to produce a more specific error.

**Comment by magnushiie** [comment]
@chilversc I'm not sure how an upcast xan help with your example. Could you show how it would be used, using the above asType<T> function (which is the equivalent to the operator I'm proposing). Note that due to parameter bivariance, the current compiler would not always give an error on invalid cast.

**Comment by chilversc** [comment]
Odd, I thought I had a case where an assignment such as `let x: Foo = {...};` would show a compile error while a cast such as `let x = <Foo> {...};` would not.

The cast was required to get the object literal to behave correctly as in this case:

```typescript
interface Foo {
    type: 'Foo',
    id: number;
}
let foo: Foo = { type: 'Foo', id: 5 };
let ids = [1, 2, 3];

//Error TS2322 Type '{ type: string; id: number; }[]' is not assignable to type 'Foo[]'.
//Type '{ type: string; id: number; }' is not assignable to type 'Foo'.
//Types of property 'type' are incompatible.
//Type 'string' is not assignable to type '"Foo"'.
let foosWithError: Foo[] = ids.map(id => ({ type: 'Foo', id: id }));

let foosNoErrorCast: Foo[] = ids.map(id => ({ type: 'Foo', id: id } as Foo));
let foosNoErrorAssignment: Foo[] = ids.map(id => {
    let f: Foo = {type: 'Foo', id: id};
    return f;
});
```

**Comment by wallverb** [comment]
Could we just use `is` the same way as `as` ?


```ts
interface A {
   a: string
}

let b = {a: 'test'} as A // type: A, OK
let c = {a: 'test', b:'test'} as A // type: A, OK
let d = {a: 'test'} is A // type: A, OK
let d = {a: 'test', b:'test'} is A // error
```

**Comment by aluanhaddad** [comment]
@wallverb that is really clever and really intuitive. Interestingly, it also provides a manifest way of describing the difference between the assignability between fresh object literals target typed by an argument vs existing objects that conform meet the requirements of that argument.

**Comment by isiahmeadows** [comment]
I like this idea of effectively a static type assertion.

**Comment by magnushiie** [comment]
@normalser your example about `let e = {a: 'test', b:'test'} is A` is to me still an up-cast and should succeed as proposed by this issue. I think your expectation is more towards #12936.

**Comment by magnushiie** [comment]
Also linking #13788 about the current behavior of the type assertion operator `as` being both upcast and downcast.

**Comment by jmagaram** [comment]
I'm new to TypeScript but have already had a need for something like this. I've been Googling for a while now and looked in the TypeScript docs and can't find what I need. I'm really surprised since I assume this capability MUST exist. I think I ran into the problem trying to use Redux and wanted to ensure that I was passing a specific type of object to the connect function. Once I specify the type hint - "this is supposed to be a X" - I'd like the editor to do intellisense and show me what properties need to be filled in, so maybe the type name needs to come first like a safe cast expression.

    function doSomething(obj: any) { }

    interface IPoint {
        x: number;
        y: number;
    }

    // This does what I expect. I'm making sure to pass a correct IPoint 
    // to the function.
    let point: IPoint = { x: 1, y: 2 };
    doSomething(point);

    // But is there a more concise way to do the above, without introducing a
    // temporary variable? Both of these compile but it isn't safe since my IPoint
    // is missing the y parameter.
    doSomething(<IPoint>{ x: 1 });
    doSomething({ x: 1 } as IPoint);

    // How about this type of syntax?
    doSomething(<IPoint!>{ x: 1 });
    doSomething(<Exact<IPoint>>{ x: 1 });
    doSomething((IPoint = { x: 1 }));
    doSomething({ x: 1 } as IPoint!); // for JSX
    doSomething({ x: 1 } is IPoint);
    doSomething({ x: 1 } implements IPoint);

**Comment by RyanCavanaugh** [comment]
@DanielRosenwasser I feel like there must be some way at this point to use mapped types or something to write something that breaks the comparability relationship in a way that it only allows subtypes through?

**Comment by jmagaram** [comment]
Actually this would work for me. I just need to create a custom function that I use wherever I need an exact type. Seems kind of obvious in retrospect. I'm still surprised this kind of thing isn't built-in.

    function exact<T>(item:T): T {
        return item;
    }

    doSomething(exact<IPoint>({x:1, y:2}));

**Comment by ackvf** [comment]
I am also looking for something like this - to specify the type of the inline object I am pushing

![image](https://user-images.githubusercontent.com/2313018/39961905-ca228666-5640-11e8-9005-786ea0cdaaa9.png)

so that when I `ctrl+space` I get hints and warnings.
By using `as` I get hints, but it is **dangerous** as I get no warnings for missing or unknown keys.

**Comment by wkrueger** [comment]
![image](https://user-images.githubusercontent.com/2209181/43441610-cc0e4392-9471-11e8-97aa-89e3e63e491f.png)

Does this belong to this issue? (Allow a way to point the type checker to strictly infer strings and numbers as constants)

**Comment by RyanCavanaugh** [comment]
Update from suggestion review - there are a bunch of scenarios where we want a syntax that looks something like

    expr SOMETHING T

Which causes `expr` to be contextually typed by `T`, but the type of the entire expression is still `expr`'s inferred type rather than `T`. An error occurs if `expr` is not assignable *to* `T` (assignability in the other direction is not sufficient).

The problem is what SOMETHING is. We punted a bunch of syntax ideas around but hated all of them. `extends` sounds too much like it does a runtime thing; `typeof` is taken; really *all* new expression syntax is super suspect, etc.. So we're basically blocked on finding some palatable syntactic space for this. Open to ideas but we're not super hopeful yet.

**Comment by magnushiie** [comment]
> Which causes expr to be contextually typed by T, but the type of the entire expression is still expr's inferred type rather than T. 
@RyanCavanaugh while I can see how this operator could be useful in some scenarios, your proposed operator is different from what I originally proposed. My proposed operator's type would be `T`, not `typeof expr`, and is useful for guiding bottom-up type inference and hiding too specific types, whereas the operator you described is mostly a documentation/debugging tool as it doesn't have any effect to the surrounding code.

**Comment by isiahmeadows** [comment]
This precisely is the minimum functionality I would like to see - we're not
concerned about contextual typing, but working around either failed type
inference or overload selection. And the easiest way to do this is to have
a safe "upcast" that's checked to not also downcast.

**Comment by ackvf** [comment]
I also see a difference in what you propose @RyanCavanaugh. In its simplest form, what I believe is desired from this proposal is that the abstract method `push` of the firebase handle gets the type assurance similar to the `push` of the typed array.

![image](https://user-images.githubusercontent.com/2313018/44512321-6ac19700-a6ba-11e8-94c9-5136c5e4d687.png)

Ideally an `is` keyword, as was already suggested, is what seems to be the best fit. `be` would be also fine.

    getFirebaseRef().push({} is MyItem)
    getFirebaseRef().push({} be MyItem)

In this example, it is like using the type guard

    const incompleteTasks = (tasks: Task[]) => tasks.filter(task => !(getWork(task.currentAssignment) is Todo).isComplete);
    const incompleteTasks = (tasks: Task[]) => tasks.filter(task => !(getWork(task.currentAssignment) be Todo).isComplete);

    // verbose, but does the same thing

    function isTodo(task: Todo | Done): task is Todo {
      return (<Todo>task).isComplete !== undefined;
    }

    const incompleteTasks = (tasks: Task[]) => tasks.filter(task => {
      const work = getWork(task.currentAssignment)
      if (isTodo(work)) {
        return !work.isComplete
      } else {
        throw new Error('Wrong Type!')
      }
    });

**Comment by mpawelski** [comment]
@RyanCavanaugh  I'm also wondering why you want `expr SOMETHING T` to be of `expr` inferred type instead of just `T`.

This still will be hugely useful feature, used quite often in many scenarios. But I don't see scenarios when I don't want the type to be `T`. 

Like in here:
![image](https://user-images.githubusercontent.com/2056282/44539356-44284e00-a704-11e8-9f36-425a8032f3a0.png)

I want this type to be `SomeInterface` not `{ a: string; b: string: c: string; }` because when I see this name in editor hints and error message I know what it is immedietely.

As for `SOMETHING`. I don't know. I kinda like `is` proposed earlier. Whatever you'll choose will be fine for me, just don't make it too long, because it will me used often 😜.

Or I'm missing the purpose of this proposal completely? Your proposal mean that `expr SOMETHING T` will throw compiler error is all cases when `var t : T = expr` throws also? right?

**Comment by RyanCavanaugh** [comment]
Scenarios where the expression type is more valuable -

```ts
// Today
const a: Partial<Point> = { x: 10 };
// Error, 'x' might be undefined
console.log(a.x.toFixed());
// OK, really shouldn't be
console.log(a.y!.toFixed());

// Desired
const a = { x: 10 } SOMETHING Partial<Point>;
// OK
console.log(a.x.toFixed());
// Error
console.log(a.y!.toFixed());

// Another example (Desired only):
type Neat = { [key: string]: boolean };
declare function fn(x: { m: boolean }): void;
const x = {
    m: true
} SOMETHING Neat;

// Today: Should be OK, isn't
fn(x);
// Today: Should be error, isn't
console.log(x.z);
```

**Comment by mpawelski** [comment]
yep, I meant that as in downcast I want new operator to be of type `T`, maybe I didn't made it clear enough.

Anyway, your examples totally convinced me why it shouldn't. thanks!

**Comment by isiahmeadows** [comment]
@RyanCavanaugh What I would like is basically syntax for this:

```ts
function is<T>(value: T): T { return value }

// Syntax for this:
is<Foo>(value) // upcast to `Foo`
```

Technically, this is a checked upcast, whereas `as` is an unchecked one (it can cast down as well as up). But this above is the semantics of any operator I'd prefer.

Just thought I'd clarify what I'm looking for.

**Comment by RyanCavanaugh** [comment]
The fact that you can write `is<T>(x): T` yourself but can't write `check<T>(x extends T): typeof x` (??) is another reason we want any hypothetical new operator to return the expression type rather than the check type.

**Comment by ORESoftware** [comment]
Yes, something like this:
https://github.com/Microsoft/TypeScript/issues/27127

**Comment by mpawelski** [comment]
>  We punted a bunch of syntax ideas around but hated all of them. 

Well, I don't know what kind of keyword team hated so I risk that I'll propose another one you'll hate 😨

I propose it to be `assertis`. I like it because it is short and has 'assert' inside and it's basically a compile-time assertion. It can only result in compile error, if there's no error then the result is the same as if we didn't use this language feature.

**Comment by simonbuchan** [comment]
Flow uses simply `:` for [a very similar feature](https://flow.org/en/docs/types/casting/) they call casting for some bizarre reason. They require a wrapping `()` to avoid ambiguity, which I think is fine in this context, since you aren't going to want to double contextually type unless you're doing something real weird like type testing.

Another precedent is [Haskell](https://www.haskell.org/hugs/pages/users_guide/type-annotations.html), which uses `::` for the same thing.

**Comment by fqborges** [comment]
I would live this implemented to use with string literals, since this is valid but obviously wrong:

    const x = { type: 'notanitem' as 'item', value: "example" }

If I have to write the string twice, at least the compiler should hint when I mess up.

    // votes for 'is'
    const x = { type: 'item' is 'item', value: "example" }
    // or flow style
    const x = { type: ('item' : 'item'), value: "example" }

**Comment by ajafff** [comment]
@fqborges what you are describing in the first part of your post is also called "sidecasting" and was marked as "working as intended" in #16995.
However, I've written a lint rule to detect this so you don't mess up: https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/no-invalid-assertion.md

**Comment by lonewarrior556** [comment]
SOMETHING: 

`mustbe`, 
`satisfies`, 
`obeys`.
`matches`,
`is`, `be`,
`compares`
`assertis`
`complements`,
`parallels`,
`achieves` ,
`realizes`,
`fulfils`

_There are only two hard things in Computer Science: cache invalidation and naming things.

-- Phil Karlton_

**Comment by DanielRosenwasser** [comment]
I like `satisfies` or `withtype`.

*TypeHintExpression* ::
&nbsp;&nbsp; *Expression* `withtype` *Type*
&nbsp;&nbsp; *Expression* `satisfying` *Type*

**Comment by eamodio** [comment]
`is` is my pref and it has a nice corollary with `as`

**Comment by RyanCavanaugh** [comment]
I wanted to set aside the syntax conversation for a moment and collect some "real-world" use cases for where this operator would be useful.

e.g.
```ts
declare function paint(color: Color): void;

export type Color = { r: number, g: number, b: number };

// All of these should be Colors, but I only use some of them here.
// Other modules are allowed to use any of these properties.
// I have no way to find the typo here:
export const Palette = {
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0},
    blue: { r: 0, g: 0, b: 255 },
} /* ??op?? Record<string, Color> */;
//          ^^^ would cause the error to be identified

paint(Palette.white);
paint(Palette.blue);
```

**Comment by jack-williams** [comment]
Heres a hacky way of doing it (error messages are poor).
```ts
export type Color = { r: number, g: number, b: number };

const check = <U>(x: U) => <T extends U extends T ? unknown : [U, 'not assignable to', T]>(): U => x;

export const Palette = check({
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0},
    blue: { r: 0, g: 0, b: 255 },
})<Record<string, Color>>();
```

**Comment by fqborges** [comment]
> 
> I wanted to set aside the syntax conversation for a moment and collect some "real-world" use cases for where this operator would be useful.
> 

My main use case that appears a lot is in my example above (https://github.com/microsoft/TypeScript/issues/7481#issuecomment-438709565).

`const x = { type: 'item', value: "example" }; // 'item' as 'item' literal not string`

What I noticed is that in the latest versions of TypeScript most of the time when I just use the above it infers the literal type.

The second use case is just hint the type on the middle of a object, without being obliged to declare the  the object as a whole. Example:

```
const myConfig = {
  name: 'foobar',
  uri: 'foo://bar',
  // ...
  // imagine a lot of props with distinct shapes here
  // ...
  palette: {
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0},
    blue: { r: 0, g: 0, b: 255 },
  } as Record<string, Color>
}
```

**Comment by matt-tingen** [comment]
Is a prefix operator an option? This would allow for intellisense to provide completions and avoid typos as you go.

**Comment by RyanCavanaugh** [comment]
> Is a prefix operator an option? This would allow for intellisense to provide completions and avoid typos as you go.

We're a bit torn on this one. You're correct that putting the type before the expression is much better for completions, but at the same time we've been pushing people toward `e as T` for a while and it seems like a natural fit to have any new operation be of the same form. But if someone finds a really good prefix syntax, that would be a big win.

**Comment by fqborges** [comment]
Going down on the prefix route, is something like this possible?

```
const myConfig = {
  name: 'foobar',
  uri: 'foo://bar',

  <Record<string, Color>>palette: {
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0},
    blue: { r: 0, g: 0, b: 255 },
  }
}
```

**Comment by matt-tingen** [comment]
Could you elaborate on the rationale for discouraging `<T>e`?

If the primary concern is with the parsing look-ahead when used with JSX, perhaps something like `<:T>e` could be used to keep the look-ahead at a constant 1 character peek. That notation is reminiscent of type annotations so it feels a bit more TypeScript-y to me.

**Comment by fqborges** [comment]
@matt-tingen If you are asking me, I am not against it. I just wrote faster than my head could elaborate my question/suggestion. 😅

**Comment by matt-tingen** [comment]
@fqborges I was meaning to ask @RyanCavanaugh. I feel like I remember seeing somewhere that the look-ahead added some complexity, but I wasn't sure if that was the main reason for encouraging `e as T` over `<T>e`. Understanding that would be beneficial if a new prefix syntax is to be considered.

**Comment by RyanCavanaugh** [comment]
`e as T` is the only one that works in TSX contests, so it's easier to just universally recommend that

**Comment by jack-williams** [comment]
Another use-case is being able to do richer excess property checking with helper types, without needing to keep the verbose helper type around:

```ts
type UnionKeys<T> = T extends unknown ? keyof T : never;
type StrictUnionHelper<T, TAll> = T extends unknown ? T & Partial<Record<Exclude<UnionKeys<TAll>, keyof T>, never>> : never;
type StrictUnion<T> = StrictUnionHelper<T, T>

const check = <U>(x: U) => <T extends U extends T ? unknown : [U, 'not assignable to', T]>(): U => x;

type Foo = {
    a : number,
    b : number,
};
type Bar = {
    a : number,
    c : number,
};

export const obj = check({
    a : 0,
    b : 0,
    c : 0,
})<StrictUnion<Foo | Bar>>(); // error
```

**Comment by fefrei** [comment]
Such an operator could also help resolve situations where type narrowing on initialization introduces a type error:

```
type A = 1 | 2;
const one: A = 1;
[one].concat([2]) // error!
```

The proposed operator allows the programmer to explicitly specify the exact type without giving up on type safety. See #36530 for details.

**Comment by w0nche0l** [comment]
I'm having a big need for something like this because a lot of Redux / React reducer hook patterns suggest to create an initial state literal, and then construct the state's type from that initial state.

For example, this is what I'm currently doing: 
```
const initItemState = {
    count: number,
    items: [] as SomeItem[],
    submissionStatus: {
        isSubmitting: false,
        errorMessage: ""
    } as GenericStatusInterface
}

type ItemState = typeof initItemState;

export function Reducer(
  state: ItemState,
  action: (actionType: string)=>ItemState
): ItemState { ... some function here}
```

However, if I later change SomeItem or GenericStatusInterface, I will not get an error because I am using `as T`.  

A small workaround that I've found that decreases legibility/increases verbosity but does give errors is doing this: 
```
const initItems: SomeItem = [];
const initSubmissionStatus : GenericStatusInterface= {
    isSubmitting: false,
    errorMessage: ""
}
const initItemState = {
    count: number,
    items: initItems,
    submissionStatus: {
        isSubmitting: false,
        errorMessage: ""
    } as GenericStatusInterface
}

```


I've only been able to casually skim the comments above, so if there has been any better workarounds for this, please let me know.

**Comment by AJGranowski** [comment]
We ran into a situation like this a little while back.

We've got this type:
```
type Thing = {
    [key: string]: Thing |  number
};
```

and this object:
```
const a = {
    foo: 4,
    bar: {
        a: 2
        b: 3
    }
};
```

We want to assert that `a` has the structure of `Thing`, but without erasing the properties of a (that is, we would want `a.c` to result in a type error).

The solution we ended up using was
```
const makeThing = <T extends Thing>(x: T): T => {
    return x;
};
const a = makeThing({ ... });
```

It would be nice to have something like `const a = { ... } is Thing` instead.

**Comment by cherryblossom000** [comment]
This would be helpful in situations like this with unions of promises of different types:
```ts
interface A { a: number }
interface B { b: number }

declare const p1: Promise<A>
declare const p2: Promise<B>

// This expression is not callable.
//   Each member of the union type '...' has signatures, but none of those signatures are
//   compatible with each other.
(Math.random() > 0.5 ? p1 : p2).then(() => {})

// Safe, but must use another variable
const safe: Promise<A | B> = Math.random() > 0.5 ? p1 : p2
safe.then(() => {})

// No extra variable/function required, but unsafe
;((Math.random() > 0.5 ? p1 : p2) as Promise<A | B>).then(() => {})
// Could fail at runtime
;((Math.random() > 0.5 ? p1 : p2) as Promise<A>).then(x => console.log(x.a.toFixed()))
```

**Comment by ff-jkrispel** [comment]
I'm working on something where this would be very useful to have. I was hoping there would be a symbol to reference the type itself, like `&` or `@` or `:` 🤷‍♂️. Just posting an example of what's in my head in case it helps:

```typescript
type StateMachine = {
  startAt: keyof &.states,
  states: {
    [key: string]: State
  }
}
```

**Comment by isiahmeadows** [comment]
@ff-jkrispel There's `this`, but that only works in interfaces. And I've used that quite a bit to do advanced type hackery with records without requiring spraying additional generic parameters everywhere.

**Comment by ThomasAribart** [comment]
@DanielRosenwasser The `is` operator described above would respond to the issue that I raised #41715. You can close it.

I'll add my use case to the list 😀

I have developed a library that infers valid types from JSON schemas only through type computations, by making profit of the `as const` statement, generics and type recursions 👉  https://github.com/ThomasAribart/json-schema-to-ts

```typescript
import { FromSchema } from "json-schema-to-ts";

const dogSchema = {
  type: "object",
  properties: {
    name: { type: "string" },
    age: { type: "integer" },
    hobbies: { type: "array", items: { type: "string" } },
  },
  required: ["name", "age"],
  additionalProperties: false,
} as const;

type Dog = FromSchema<typeof dogSchema>;
// => { name: string; age: number; hobbies?: string[] }
```

The `as const` statement is needed to keep `typeof dogSchema` as narrow (and thus parsable) as possible.

Currently, `FromSchema` will raise an error if `typeof dogSchema` doesn't extend DefinitelyTyped's JSON schema definitions. It's better than nothing, but it would be great to have the benefits of a type assignment (autocomplete, error highlighting etc...).

```typescript
import { JSONSchema } from "json-schema-to-ts"; // <= Readonly type based on DefinitelyTyped

const invalidSchema = {
  type: "object",
  properties: {
    name: "string" // <= Error is raised
} as const is JSONSchema;

const validSchema = {
  type: "object",
  properties: {
    name: { type: "string" },
  }
} as const is JSONSchema;

type ValidData = FromSchema<typeof validSchema> // <= Still works
```

**Comment by sschultze** [comment]
This would be very helpful. I'm currently using a helper function

```
function isType<T>(value: T): T {
  return value;
}
```

in my source code, but this obviously creates an unnecessary function call.

**Comment by RodrigoRoaRodriguez** [comment]
Out of the top of my head it occurs to me that there are at least four scenarios where contextual types would be beneficial to Typescript:
- Objects
- Arrays
- Generics
- Functions

When I wrote #42138, a related issue, it helped to write proposals in order to illustrate what I meant.  I will do the same here.

# Proposals
These proposals all specify contextual types using the keyword `infer` used outside of conditional types. I will use `string` for all non-contextual types provided by the user.
  
## Objects
```ts
// LHS, Infer keys types
type ValuesMustBeStrings = {[key: infer] :  string}
// RHS, Infer value types
type KeysMustBeStrings = {[key: string] :  infer}
```
This would be very useful for declaring collections objects with constrains that need to be followed. The two main work arounds for this are index types, or `as const` which are not meant to be used used for this use-case and have undesirable side-effects (e.g. forcing you to hardcode keys and values and making everything readonly respectively). 

## Arrays
```ts
// Infer value types
type LengthTwoTuple = [infer, infer]
type  FirstValueMustBeAString = [strings, …infer]
```
This would be very useful for tuple types, as [the current way of typing an inferred tuple is through hacky typing-functions](https://stackoverflow.com/questions/49729550/implicitly-create-a-tuple-in-typescript).

## Generics
```ts
// Assume a type from a third party library with three generic argument A,B and C
const myArglessThing: TypeFromLibrary =  new Thing();
const myExplicitlyArglessThing: TypeFromLibrary<infer, infer, infer> = new Thing();
//Assume now that the default values are unsuitable and I don’t want override inference of the last two types.  
// Either because I don’t want to or can’t easily do so(say the default types for the the second two values are private)
const myThing: TypeFromLibrary<string, infer, infer> =  new Thing(’lets say this is the string value’);
```

## Functions 
```ts
type Both_inferred = infer => infer // this would be the explicit equivalent to not providing a type on declaration
type Lhs_inferred = infer => number // only the argument is inferred
type Rhs_inferred = ( arg: number ) => infer // only the return type is inferred
```
See: #42138 where I wrote extensively about this functions with contextual typings.

**Comment by sschultze** [comment]
@RodrigoRoaRodriguez I have read your comment twice, I don't have the feeling that it has anything to do with this issue. Perhaps you posted it at the wrong place.

**Comment by RodrigoRoaRodriguez** [comment]
Others seem to think #42243  is related to #7481. While superficially the two issues might seem unrelated. There is a commonality: the workaround.  Look at the my examples of function workarounds, and you will realize that they are all either variants of, or identical to:
```ts
function asType<T>(value: T) {
  return value;
};
```
In fact, my own example was also component actions just that instead of an operator:
```ts
export const ConnectedAsType = connect(undefined, asType<MyComponentActions>({
  selectSomething,
}))(MyComponent);
``` 
My suggestion was a Type : 
```ts
export const ConnectedAsType: ConnectedComponentActions = connect(undefined, {
  selectSomething,
}))(MyComponent);
```
[Bear in mind that for my example of state management I used `useReducer` rather than `Redux`](https://codesandbox.io/s/why-typescript-would-benefit-from-partially-typed-functions-td0xu?file=/src/ReducerState.tsx). However, the core intent is the same and both seek achieve the same result. Just through different methods.

**Comment by btoo** [comment]
looks like @ThomasAribart [beat me](https://github.com/microsoft/TypeScript/issues/7481#issuecomment-736103803) to [my own suggestion](https://github.com/microsoft/TypeScript/issues/42234). that being said, if i may offer one suggestion on top of his - I think there's a big opportunity here to re-use the `implements` keyword rather than creating a new `is` operator:
```
const validSchema = {
  // ...
} as const implements JSONSchema;
```

**Comment by jtlapp** [comment]
I recently posted [another use case](https://stackoverflow.com/questions/69957133/construct-object-with-new-type-conforming-to-another-type?noredirect=1&lq=1) to StackOverflow. I'm generating an API from a structure and need the structure to have a certain conformation while still preserving its unique type.

**Comment by dhoulb** [comment]
Most of the time I tend to do something like this:

```ts
export const STRING_CONVERT = {
    num: (num: number) => num.toString(),
    bool: (bool: boolean) => (bool ? "yes" : "no"),
};

// Check that STRING_CONVERT only includes functions that return strings.
STRING_CONVERT as Record<string, (v: unknown) => string>;
```

I know this leaves a random `FUNCTIONS;` in the code but tree shaking cleans it up (maybe?). Flow has assertion syntax (using `:`) that _does_ get removed from output so I started doing something similar when I switched to TS.

But I agree the following with `implements` is much cleaner (and a great suggestion!):

```ts
export const STRING_CONVERT = {
    a: (num: number) => num.toString(),
    b: (bool: boolean) => (bool ? "yes" : "no"),
} implements Record<string, (v: unknown) => string>;
```

**Comment by noppa** [comment]
@dhoulb Your example highlights why `as` and its downcasting behaviour is bad at doing assertions like that

```typescript
export const STRING_CONVERT = {
    // woops, the actual return type here is 0 | string
    num: (num: number) => num && num.toString(),
};

// Passes type check without errors :(
STRING_CONVERT as Record<string, (v: unknown) => string>;
```
whereas
```typescript
const _check: Record<string, (v: unknown) => string> = STRING_CONVERT;
```
would give you an error as expected.

**Comment by voliva** [comment]
> I recently posted [another use case](https://stackoverflow.com/questions/69957133/construct-object-with-new-type-conforming-to-another-type?noredirect=1&lq=1) to StackOverflow. I'm generating an API from a structure and need the structure to have a certain conformation while still preserving its unique type.
> 
> All I'd like is a declarative shorthand for the following function:
> 
> ```ts
> function assertTemplate<T extends Template>(template: T) {
>   return template;
> }
> ```
> 
> Maybe something like:
> 
> ```ts
> const api: ConformsTo<Template> = {
>     // ...
> };
> ```
> 
> Which would be equivalent to:
> 
> ```ts
> const api = assertTemplate({
>     // ...
> });
> ```

I usually have a use case very similar to @jtlapp, but I think it's specially relevant when there are generics in between.

Simplified example, I have an interface with a generic:

```ts
interface ColumnDefinition<M, T> {
  id: string,
  selector: (model: M) => T,
  renderer: (value: T) => JSX.Element
}
```

If I want to instantiate a variable with this type, I have to put the generics in:

```ts
const labelColumn: ColumnDefinition<Todo, string> = {
  id: 'label'
  selector: todo => todo.label,
  renderer: label => ...
}
```

and I can't avoid it directly. If I put in `any` for the generics (or default them), then `renderer` will have "label" as any too. I usually have the same identity function that does the inference for me:

```ts
function colDef<M, T>(columnDefinition: ColumnDefinition<M, T>) {
  return columnDefinition
}

const labelColumn = colDef({
  id: 'label'
  selector: (todo: Todo) => todo.label,
  renderer: label => ... // T inferred nicely as string
})
```

**Comment by jtlapp** [comment]
I have a situation for which I haven't found a workaround. I want each element of an array to conform to the same abstract structure without requiring all the elements to be of the same type. I can't use classes because I'm dictating that all object members have a certain structure, regardless of their names.

It seems that the `ConformsTo<T>` proposal would address this:

```ts
function processRecords(records: ConformsTo<RecordStructure>[]) {
    ...
}
```

The problem is that each element of the array can have a different set of properties.

**Comment by ethanresnick** [comment]
I think #33480 could also address many of the use cases here

**Comment by rpeszek** [comment]
I would like to propose a workaround,  It works in current TypeScript v4.5.2.

```TypeScript
type Same<P,T> = P extends T? (T extends P? 1: 0): 0

const verifySame = <P> () => <T> (_: Same<P,T>, t:T): T => t

type Hello = {hello: string}

verifySame<Hello>()(1, {hello: "world"})
verifySame<Hello>()(0, {hello: "world", since : 2020})

//this will not compile:
//Argument of type '1' is not assignable to parameter of type '0'.ts(2345)
//verifySame<Hello>()(1, {hello: "world", since : 2020})
```

`verifySame` has a  higher rank type but appears to be supported by TS. 
verifySame should provide type safety preventing subtyping. 

I am working on next installment of my "Type Enthusiast's Notes about TypeScript" and have been experimenting with 
things like higher rank and existential type constructions.   A lot of these concepts translate to TS and together with some type level programming can yield interesting type safety.

**Comment by rpeszek** [comment]
To clarify my comment, the rank-2 is convenient in some case but often not needed.  Cleaning up my above example (replacing literal numbers with literal booleans) here is a safe invariant version of the array `push`:

```TypeScript
//Cleaner version of the above 'Some' uses literal booleans
type Same<P,T> = P extends T? (T extends P? true: false): false

//Note to get safety I end up with casting, this is quick and dirty but still ...
const safePush = <P, T> (_: Same<P,T>, ps: P[], t: T): number => ps.push(t as any)

const intlist: number[] = [1,2,3]
const list: unknown[] = intlist
list.push("not a number")  //unsafe TS push adds 'string' to 'intlist'

safePush(true, intlist, 1)  //this is safe
//safePush(true, list, 1)    //this is risky and will not compile
//safePush(true, list, "not a number") //this is risky (here wrong) and will not compile
```

This does not use rank-2 tricks and nicely fixes variance problem.  It shows how such work-around could  be used in APIs.

**Comment by RyanCavanaugh** [comment]
I have some open questions based on playing with the PR and would like to hear people's thoughts
```ts
/* ~~ Example 1 ~~ */
const x = [1, 2] satisfies [number, number];
// What is the type of 'x' ?
// Argument for [number, number]:
//   If [1, 2] isn't a [number, number], then the 'satisfies' should have been rejected, but it wasn't
// Argument for number[]:
//   This operator is not supposed to change the type of the expression (except to remove implicit anys)

/* ~~ Example 2 ~~ */
type Disposable = {
    dispose(): void;
    beforeDispose?(): void;
};
const s = {
    init() {
        // Allocate
    },
    dispose() {
        // Cleanup
    }
} satisfies Disposable;
// Is 'init' an excess property here?
// Argument for:
//   If instead of 'init' we had written 'beforDispose' [sic],
//   failing to flag that as an error is a clear deficit
// Argument against:
//   This code sure looks fine. You could have written
//    } satisfies Disposable & Record<string, unknown>
//   instead if you were OK with excess properties

/* ~~ Example 3 ~~ */
let y = 3 satisfies 3;
let z: 3 = y;
// Is this a correct error?
//   Argument for: This operator is not supposed to change the types. You could have written 'as const' or 'as 3' instead.
//   Argument against: I literally just checked that this is 3!
```

**Comment by RyanCavanaugh** [comment]
> `const foo = null satisfies string | null;`

Based on prior discussion, I don't think we want `foo` to be `string | null` here. The reason is that, using the example from my prior comment, you want to be able to call `s.init()` -- this implies we should use the expression type, not the asserted type.

**Comment by fatcerberus** [comment]
I was assuming that `init` in that example was going to be treated as an excess property, consistent with `satisfies` being a contextual-typing hint.

**Comment by fatcerberus** [comment]
I guess I find any solution where `satisfies` can’t change the type of an expression to be potentially confusing, since the type of an expression is usually affected by contextual typing anyway.  It just seems natural that “the contextual typing operator” should be able to change the type of an expression.  It’s more mental overhead if I have to remember that `satisfies` contextually types an expression, but doesn’t affect its type like other cases of contextual typing do.  TypeScript IMO has way too many of these special cases already.

**Comment by noppa** [comment]
> // What is the type of 'x' ?

If the type of `x` is not affected by `satisfies` and `[1, 2] satisfies [number, number]` becomes `number[]` again, then I guess you loose the ability to use this feature to annotate individual properties of objects? Expression level type annotations could be very handy to avoid boilerplate when you have lots of properties on an object and would rather not repeat them in the type definition and its implementation. I.e. I'd like the following to error, [as it currently does in the playground](https://www.staging-typescript.org/play?ts=4.6.0-pr-46827-10#code/GYVwdgxgLglg9mABBATgUwIZTQFQBYxgDmAFAM5QoBciFKhRAlIgN4BQini6UIKS7LkMTA4cGgG06AXVpYYZYDDRlEUyg2kAaDlwC+bA21SZs+BiQDkouJcYA6GxICM0+1DgBVAA7e0KAGEMMjQSRjYAegjhGNiuAD1EAGU8OBAAGwATRDA4KEQAIzREDHT0uAB3NGyPEogIFVVCTLQAD0RnRDhgNTpNIA)

```typescript
function createThing(str: string) {
    return {
        foo: [str] satisfies [string],
    }
}

createThing('foo').foo[1].toUpperCase()
//                     ^ Should not be allowed to access index 1 of [string]
```
> // Is 'init' an excess property here?

Yes please. Guarding against excess properties (which are often typos, leftovers from old APIs or other mistakes) is a common need and being able to do it inside arbitrary expressions would be very handy.

Basically I was hoping just expression-level `: `, which I use quite often in Flow.   
I guess this directly conflicts with @jtlapp 's use cases though. 

> // Is this a correct error?

Seems weird to me.  
Why `x = [1, 2] satisfies [number, number]` is `[number, number]` but `3 satisfies 3` is not `3`? :thinking:

**Comment by Harpush** [comment]
I hope that's what this feature means - but a good scenario is combining `as const` with type annotation. Currently there is no way to say `const a = {a: 2} as const satisfies Record<string, number>` which in this case will make sure the object follows the record structure but at the end the type will be the actual const asserted type.

**Comment by noppa** [comment]
@jtlapp If I understand your use-case correctly, you can already do something like
```typescript
type ApiTemplate = {
  [funcName: string]: {
    call: (data: any) => void;
    handler: (data: any) => void;
  }
};

const apiConfig1 = () => {
  return {
    func1: {
      call: (count: number) => console.log("count", count),
      handler: (data: any) => console.log(data),
    },
    func2: {
      call: (name: string) => console.log(name),
      handler: (data: any) => console.log("string", data),
    },
  }
};

// Currently working way to do type validation without any runtime impact
const _typecheck: ApiTemplate = null! as ReturnType<typeof apiConfig1>
```
[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAgmCWAVCBbMAbAhsaBeKA3gFBRQDaAZgK4B2AxgHKYoQBcUAzsAE7w0DmAXXbFSpOpnTp2ACgAm2TO0w0QASii4AfFABuAe3hyA3CTEALFXPQRushcCVQV6zToNHTpAL5Fvpojp9Gi5nBABhYIp4fgBGTSgZDW1CM24IYCpuGlSxKGp6WJEzPIkpWSDaYHYaKhQAI1tknSCQ-RsAOnR9fhkAIkqaYD6AGihB4DURkosrGztEhycXZvHgjnaILp75RSmZ72m8groAJmK88UlpRJpmNk4ePn5V1o3O7t67ln3LqEsaNZbPZFMpVK91pttr0+lxeAJRlAlr8xIczL5-EQiAB6bFQcJZdJDdAgKAAd303AA1s9yZhScB9EimaBIHpJEZsPBguT4MBzPoqMBnKooNwqvAWFBJWBMHRgIF1sKAPqsiB0czqqnsOBIVAYbB4KC1KQAQmcHCgACUMlkaIhwBAADxq-QUMLwSI0aJxLREIA).

The unfortunate thing about that approach is that it involves using an extra unused variable, which tends to make linters mad. `satisfies` would fix this, as you could instead do
```typescript
// Assert that apiConfig returns a valid ApiTemplate
null! as ReturnType<typeof apiConfig1> satisfies ApiTemplate
```
[Playground](https://www.staging-typescript.org/play?ts=4.6.0-pr-46827-10&ssl=23&ssc=1&pln=20&pc=1#code/C4TwDgpgBAgmCWAVCBbMAbAhsaBeKA3gFBRQDaAZgK4B2AxgHKYoQBcUAzsAE7w0DmAXXbFSpOpnTp2ACgAm2TO0w0QASii4AfFABuAe3hyA3CTEALFXPQRushcCVQV6zToNHTpAL5Fvpojp9Gi5nBABhYIp4fgBGTSgZDW1CM24IYCpuGlSxKGp6WJEzPIkpWSDaYHYaKhQAI1tknSCQ-RsAOnR9fhkAIkqaYD6AGihB4DURkosrGztEhycXZvHgjnaILp75RSmZ72m8groAJmK88UlpRJpmNk4ePn5V1o3O7t67ln3LqEsaNZbPZFMpVK91pttr0+lxeAJRlAlr8xIczL5-EQiAB6bGwDgcWzAKDASzEzARKIxKDpTLZDjOPSSIywBDINBYHBEWpSACEzgZACUMlkaIhwBAADygSD6ChheCRGjROI6DjYeAcaIQBlwJCoDDYCBY3FQcJZdJDdAgKAAd303AA1s87ZgbcB9EjPTLoLpmQ54ME7fBSfoqOTVDSqvAWFAY2BMHRgIF1sSAPo+ujmCB0R3sPXsw04BI89D8zBCkXZcWQaUSuUKpUq2JaIhAA)

**Comment by jtlapp** [comment]
Thank you, @noppa. I actually do have my code working, but it's missing an important piece, which it seems that I didn't overtly say in my use cases above:

_I want VSCode highlighting the errors in the user's custom ApiTemplate definition, rather than at a later point in the code where the user passes that custom definition to some function._

The goal is to facilitate correct API development at the point where the API code is written.

I'm close to sharing my project to provide better clarity on the problem I'd like to solve. I just need to document it. I've found the Electron IPC mechanism so-many-ways error prone, and I've developed a way to create IPCs merely by creating classes of asynchronous methods, registering these classes on one end, and binding to them on the other. The entire objective is to make life as easy and bug-free as possible for the developer. I'm actually pretty amazed that I was able to do it. The only thing lacking at this point is having VSCode indicate API signature errors in the API code itself. I used to think the TypeScript type system was excessively complicated, but now I'm addicted to its power.

**Comment by jmroon** [comment]
The only way satisfies is useful to me is if it does provide additional context to the inferred type. 

Use case: providing more accurate types for object literal properties. This is particularly useful in something like Vue Options API where everything is declared in an object literal.

**Comment by treybrisbane** [comment]
@RyanCavanaugh I took a look at your questions above, and these are my thoughts:

---
### Example 1
> ```ts
> const x = [1, 2] satisfies [number, number];
> // What is the type of 'x' ?
> // Argument for [number, number]:
> //   If [1, 2] isn't a [number, number], then the 'satisfies' should have been rejected, but it wasn't
> // Argument for number[]:
> //   This operator is not supposed to change the type of the expression (except to remove implicit anys)
> ```
In this case, I'd expect the type of `x` to be `[number, number]`, for two reasons:
1. The type of the expression `[1, 2]` is `number[]`, but since `[1, 2]` is assignable to type `[number, number]`, _and_ since `[number, number]` is more specific than `number[]`, I expect the type of the overall expression to be `[number, number]`
2. My mental model of the `satisfies` operator is essentially "assume _at least_ this shape if (and only if) it's safe to do so", rather than "verify that it would be possible to assume this shape"

---
### Example 2
> ```ts
> type Disposable = {
>     dispose(): void;
>     beforeDispose?(): void;
> };
> const s = {
>     init() {
>         // Allocate
>     },
>     dispose() {
>         // Cleanup
>     }
> } satisfies Disposable;
> // Is 'init' an excess property here?
> // Argument for:
> //   If instead of 'init' we had written 'beforDispose' [sic],
> //   failing to flag that as an error is a clear deficit
> // Argument against:
> //   This code sure looks fine. You could have written
> //    } satisfies Disposable & Record<string, unknown>
> //   instead if you were OK with excess properties
> ```
This is a tricky one.

At first I was going to make the argument that this case _should_ raise an excess property error specifically because this code does:
```ts
const s: Disposable = {
  init() { ... },
  dispose() { ... }
};
```
However, I realised that such behaviour would contradict my mental model of the operator.

I mentioned above that my mental model of `satisfies` was essentially "assume _at least_ this shape if (and only if) it's safe to do so". That "at least" bit is important, because it means that if the LHS is of a _more specific_ type than the RHS (while still being assignable to the RHS), then I expect the overall type of the `satisfies` expression to be that of the LHS, _not_ the RHS.
For example:
```ts
const value1 = (3 as const) satisfies number; // `value1` should be of type `3`
const value2 = ([1, 2, 3] as [1, 2, 3]) satisfies number[]; // `value2` should be of type `[1, 2, 3]`
```

So, revisiting the example from the question with this in mind:
```ts
type Disposable = {
  dispose(): void;
  beforeDispose?(): void;
};

const s = {
  init() { ... },
  dispose() { ... }
} satisfies Disposable;
```
I would expect the type of `s` to be
```ts
{
  init(): void;
  dispose(): void;
}
```
and _because_ the type of `s` is not _specifically_ `Disposable`, I would not expect to see an excess property error.

---
### Example 3
> ```ts
> let y = 3 satisfies 3;
> let z: 3 = y;
> // Is this a correct error?
> //   Argument for: This operator is not supposed to change the types. You could have written 'as const' or 'as 3' instead.
> //   Argument against: I literally just checked that this is 3!
> ```
I feel like this should _not_ be an error.
My interpretation of this code is basically the same as the first example. The type of the expression `3` is `number`, but since `3` is assignable to type `3`, _and_ since `3` is more specific than `number`, I expect the type of the overall expression to be `3`. As a result, I expect `y` to be of type `3`, and thus its assignment to `z` to succeed.

---
I hope that helps! 🙂 

Also, please let me know if my mental model of `satisfies` is incorrect in any way, as if it is, most of what I've said here will probably be invalid! 😁

**Comment by RyanCavanaugh** [comment]
> Also, please let me know if my mental model of satisfies is incorrect in any way

Right now there are no wrong answers -- the point of the exercise to determine what the "right" mental model of this operator should be!

**Comment by simonbuchan** [comment]
I would expect something to exactly like inference for generic constrained arguments, if that helps. If there's an obviously wrong behavior for one then it's probably also wrong for the other.

**Comment by kasperpeulen** [comment]
@simonbuchan I agree. I guess that you could model the spec to this function:

```ts
export function satisfies<A>() {
  return <T extends A>(x: T) => x;
}
```

If we then look @RyanCavanaugh examples, then it corresponds with @treybrisbane mental model:

```ts
/* ~~ Example 1 ~~ */
const x = satisfies<[number, number]>()([1, 2]);
// type is inferred as [number, number]

/* ~~ Example 2 ~~ */
type Disposable = {
  dispose(): void;
  beforeDispose?(): void;
};
const s = satisfies<Disposable>()({
  init() {
    // Allocate
  },
  dispose() {
    // Cleanup
  },
});
// Is 'init' an excess property here?
// No error here. And the type is inferred as: { init(): void, dispose(): void }

/* ~~ Example 3 ~~ */
let y = satisfies<3>()(3);
let z: 3 = y;
// There is no error, type of y is inferred as 3
```

Also the other examples:

```ts
const value1 = satisfies<number>()(3 as const); // `value1` is of type `3`
const value2 = satisfies<number[]>()([1, 2, 3] as [1, 2, 3]); // `value2` is of type `[1, 2, 3]`

// and another one
const value3 = satisfies<ReadonlyArray<number>>()([1, 2, 3] as const); // `value3` is of type `readonly [1, 2, 3]`
```

**Comment by simonbuchan** [comment]
@kasperpeulen thanks for the code and examples! Glad to see it works out, I wasn't on a computer and couldn't figure out what the satisfies function would look like.

**Comment by shicks** [comment]
I'd like to interject a slight tangent (pessimistically, scope creep?) that I haven't seen pointed out yet, but I think would be a good use case for any keyword introduced for these static checks.  Specifically, some generic constraints are impossible to express due to circular references, and unlike top-level variable/expression type constraints, there's no other place to even put a dead assignment to get the check.

This has come up for me a number of times as a framework author wanting to type my APIs as tightly as possible to prevent accidental misuse.  Most recently, I was trying to write an "isEnum" type guard (see also #30611):
```
function isEnum<T extends Enum<T>>(enumContainer: T): (arg: unknown) => arg is T[keyof T] { ... }
type Enum<T, E = T[keyof T]> =
    [E] extends [string] ? (string extends E ? never : unknown) :
    [E] extends [number] ? (
        true extends ({[key: number]: true} & {[P in E]: false})[number] ?
            unknown : never) :
    never;
```
Unfortunately, because of the `[P in E]` index required to work around #26362, TypeScript sees a circular reference and can't handle the `T extends Enum<T>`, but a human can look at this and see that this is not so much a type _bound_ (that needs to be resolved in order to typecheck the body of the function), but rather a type _constraint_ that only matters for checking callsites.  This seems like an excellent use case for `function isEnum<T satisfies Enum<T>>` or `function isEnum<T implements Enum<T>>`, which would check the constraint for callers, but not have a problem with circularity.

The reason this is a good fit is that there's no other good place to do this check, due to how TypeScript does its type checking.  In C++, you can write a `static_assert` in the body of the function, and since it re-typechecks each generic function with every new instantiation, it works.  But TypeScript uses the template bounds to type check the function body once, and from then on callers are reduced to just checking the template bounds - so this allows more extensive verification that the types passed into the templates are actually reasonable.

FWIW, my current workaround is pretty ugly and doesn't work universally.  It essentially boils down to
```
type NotAnEnum = {__expected_an_enum_type_but_got__: T};
type EnumGuard<T> = ... ? (arg: unknown) => arg is T[keyof T] : NotAnEnum<T>;
function isEnum<T>(enumContainer: T): EnumGuard<T> { ... }
```
This gets the job done, but as folks upthread have pointed out, it moves the error from the construction to the usage - because the `isEnum` function typechecks no matter what, it just returns an unusable value.  This also relies on there being usable and unusable values: if the function returned `void`, for instance, this wouldn't work so well.

**Comment by kasperpeulen** [comment]
One other reason in favour of allowing `satisfies` to "assume *at least* this shape if (and only if) it's safe to do so".

When working with a lot of tuples, I often have the following problem:

```ts
declare function getTuple(): [number, number];
declare function calculate(tuple: [number, number]): number;
declare function memo<T>(factory: () => T): T;

function foo() {
  const [a, b] = getTuple();
  const tuple = memo(() => [b, a * 2]);
  // TS2345: Argument of type 'number[]' is not assignable to parameter of type '[number, number]'.
  // Target requires 2 element(s) but source may have fewer.
  return calculate(tuple); 
}
```

I think most people would first `as const`, but that gives another problem:

```ts
function foo() {
  const [a, b] = getTuple();
  const tuple = memo(() => [b, a * 2] as const);
  // TS2345: Argument of type 'readonly [number, number]' is not assignable to parameter of type '[number, number]'.   
  // The type 'readonly [number, number]' is 'readonly' and cannot be assigned to the mutable type '[number, number]'.
  return calculate(tuple); 
}
```

So now, what do you do? You have two options:

* one is to change the parameter of `calculate` to be a `readonly [number, number]`, which is fine, but also feels a bit arbitrary if you are not using readonly types in the rest of the project, and sometimes it means that many functions need to be refactored to readonly.

* the other option is, to not use `as const`, but use `as [number, number]`. This seems like a solid solution, but it is actually quite dangerous, because what if now a refactor made getTuple to return `declare function getTuple(): [number, number | undefined];` You may think that you could safely do this refactor in strict typescript, while my code stays sound, but no!

```ts
declare function getTuple(): [number, number | undefined];
declare function memo<T>(factory: () => T): T;
declare function calculate(tuple: [number, number]): number;

function foo() {
  const [a, b] = getTuple();
  // No error, but this code is not sounds anymore, silently converting number | undefined to a number
  const tuple = memo(() => [b, a * 2] as [number, number]);
  return calculate(tuple);
}
```

With `satisfies`, this could be done in a sound way:

```ts
declare function getTuple(): [number, number | undefined];
declare function memo<T>(factory: () => T): T;
declare function calculate(tuple: [number, number]): number;

function foo() {
  const [a, b] = getTuple();
  // This will not compile. The type need to be changed to [number | undefined, number] or the null value must be handled.
  const tuple = memo(() => [b, a * 2] satisfies [number, number]);
  return calculate(tuple);
}
```

**Comment by mbrowne** [comment]
Here's another use case to consider:

```ts
import type { RequestHandler } from 'express'

interface RouteHandlers {
    login: RequestHandler
    logout: RequestHandler
}

export const routeHandlers: RouteHandlers = {
    login(req, res) {
        ...
    },

    logout(req, res, next) {
        ...
    }
}
```

`satisfies` would certainly make this more DRY:

```ts
export const routeHandlers = {
    login(req, res) {
        ...
    } satisfies RequestHandler,

    logout(req, res, next) {
        ...
    } satisfies RequestHandler
}
```

...although in this case I can envision some syntax that would be even more concise, given that every value in the object should be a function of type `RequestHandler`. Obviously there are more important things than just being concise, just wanted to share the use case.

**Comment by cefn** [comment]
To summarise what I recorded in the other issue (before I was helpfully referred to this feature discussion) and responding to @RyanCavanaugh ....

> What is the type of 'x'

What I expected from a `like` operator  - it would do nothing at all to the types of any existing language construction, but it WOULD raise compiler errors if the item cannot be known to satisfy the definition.

So for cases where the type IS satisfied it's a passthrough, and nobody has to reason about it at all, while if it ISN'T satisfied, then it would raise errors just like an assignment to a type (including excess property checks for literals) giving local and immediate feedback that it wasn't `like` the type you asked.

**Comment by RyanCavanaugh** [comment]
I'd like to pick up the conversation at #47920 to get a clean comment slate. There's a large write-up there.
